File: models/Gpt-3.5-turbo/py/code/range_queries/Forest_Queries.py

# read input
n, q = map(int, input().split())
forest = [input() for _ in range(n)]
queries = [list(map(int, input().split())) for _ in range(q)]

# calculate the number of trees in each prefix rectangle
prefix_trees = [[0] * (n+1) for _ in range(n+1)]
for i in range(1, n+1):
    for j in range(1, n+1):
        prefix_trees[i][j] = prefix_trees[i-1][j] + prefix_trees[i][j-1] - prefix_trees[i-1][j-1]
        if forest[i-1][j-1] == '*':
            prefix_trees[i][j] += 1

# process queries
for query in queries:
    y1, x1, y2, x2 = query
    trees_in_query = prefix_trees[y2][x2] - prefix_trees[y1-1][x2] - prefix_trees[y2][x1-1] + prefix_trees[y1-1][x1-1]
    print(trees_in_query)

File: models/Gpt-3.5-turbo/cpp/code/problems/range_queries/Forest_Queries.cpp

#include <iostream>
#include <vector>

using namespace std;

int main() {
    int n, q;
    cin >> n >> q;

    vector<vector<int>> forest(n+1, vector<int>(n+1, 0));

    // Read forest grid
    for (int i = 1; i <= n; i++) {
        string row;
        cin >> row;
        for (int j = 1; j <= n; j++) {
            forest[i][j] = (row[j-1] == '*') ? 1 : 0;
        }
    }

    // Calculate prefix sum grid
    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= n; j++) {
            forest[i][j] += forest[i-1][j] + forest[i][j-1] - forest[i-1][j-1];
        }
    }

    // Process queries
    for (int query = 0; query < q; query++) {
        int y1, x1, y2, x2;
        cin >> y1 >> x1 >> y2 >> x2;

        int treesInsideRectangle = forest[y2][x2] - forest[y2][x1-1] - forest[y1-1][x2] + forest[y1-1][x1-1];
        cout << treesInsideRectangle << endl;
    }

    return 0;
}
